File: days/2/js/day-2.js
const fs = require('node:fs/promises');

async function main() {
    const input = await fs.readFile('days/2/input-2.txt', 'utf8');

    const lines = input.split('\n');

    let safeReports = lines.length;

    for (const line of lines) {
        const lineNumbers = line.split(' ').map(Number);

        let trend = null;
        for (let i = 1; i < lineNumbers.length; i++) {
            const prevNum = lineNumbers[i - 1];
            const num = lineNumbers[i];
            const diff = Math.abs(prevNum - num);

            if (diff < 1 || diff > 3) {
                safeReports -= 1;
                break;
            }
            const sign = prevNum > num ? -1 : 1;
            if (trend === null) {
                trend = sign;
            } else if (trend !== sign) {
                safeReports -= 1;
                break;
            }
        }
    }

    console.log('Safe reports num -', safeReports);
}

main();

File: days/2/go/day-2.go
package main

import (
	"fmt"
	"os"
	"path/filepath"
	"strings"
	"strconv"
)

func partOne() int {
	absPath, _ := filepath.Abs("days/2/input-2.txt")

	inputData, err := os.ReadFile(absPath)
	if err != nil {
		panic(err)
	}

	lines := strings.Split(string(inputData), "\n")

	var safeReports int = 0

	for _, line := range lines {
		parts := strings.Split(line, " ")

		trend := 0
		for i := 1; i < len(parts); i++ {
			prevNum, _ := strconv.Atoi(parts[i-1])
			num, _ := strconv.Atoi(parts[i])
			
			diff := prevNum - num

			currentTrend := 1
			if diff < 0 {
				currentTrend = -1
			}

			if trend == 0 {
				trend = currentTrend
			} else if trend != currentTrend {
				safeReports -= 1
				break
			}

			if absDiff := diff * currentTrend; absDiff < 1 || absDiff > 3 {
				safeReports -= 1
				break
			}
		}
		safeReports += 1
	}

	return safeReports
}

func main() {
	safeReports := partOne()

	fmt.Printf("[Part 1] Safe reports num - %d \n", safeReports)
}